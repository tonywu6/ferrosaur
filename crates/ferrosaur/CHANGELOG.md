# CHANGELOG

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and
this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.1.1](https://github.com/tonywu6/ferrosaur/compare/ferrosaur-v0.1.0...ferrosaur-v0.1.1)

### Other changes

#### <!-- 3 --> Documentation

- avoid layout shift due to badges [`b64b5b3`](https://github.com/tonywu6/ferrosaur/commit/b64b5b3faa1f9ee61cba4ad149dd589356681fb9)
- reword README [`1134a63`](https://github.com/tonywu6/ferrosaur/commit/1134a63cf92e52d7c62607818505e1a24a0a9cc2)
- adjust examples [`9eb54f5`](https://github.com/tonywu6/ferrosaur/commit/9eb54f5b91df013cb1e8c1c75bbde6fdbeb691ba)

## [0.1.0](https://github.com/tonywu6/ferrosaur/releases/tag/ferrosaur-v0.1.0) - 2025-05-07

Initial release!
