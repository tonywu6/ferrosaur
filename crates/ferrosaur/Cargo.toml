[package]
name = "ferrosaur"
publish = true
readme = "../../README.md"
version = "0.1.0"

authors.workspace = true
edition.workspace = true
license.workspace = true
repository.workspace = true

[lib]
proc-macro = true

[dependencies]
darling = "0.20.10"
heck = "0.5.0"
proc-macro2 = { workspace = true }
quote = { workspace = true }
syn = { workspace = true }
tap = { workspace = true }

# only for docs
deno_core = { workspace = true, optional = true }

[dev-dependencies]
anyhow = { workspace = true }
deno_console = { workspace = true }
deno_core = { workspace = true }
deno_url = { workspace = true }
deno_web = { workspace = true }
deno_webidl = { workspace = true }
insta = { workspace = true }
serde = { workspace = true }
serde_json = { workspace = true }
tokio = { workspace = true }

[features]
_unstable-docs = ["dep:deno_core"]

[package.metadata.docs.rs]
features = ["_unstable-docs"]
