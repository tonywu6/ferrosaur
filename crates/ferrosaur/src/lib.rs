use darling::{
    ast::NestedMeta,
    util::{path_to_string, Flag},
    Error, FromMeta, Result,
};
use proc_macro2::TokenStream;
use syn::{parse_macro_input, Lit, LitStr, Meta};

mod callable;
mod fast_string;
mod global_this;
mod interface;
mod iterator;
mod module;
mod util;
mod value;

use crate::util::{
    flag::{FlagEnum, FlagError, FlagLike, FlagName},
    positional::Positional,
    property::WellKnown,
    string::StringLike,
    unary::Unary,
    v8::V8InnerType,
    FatalErrors, TokenStreamResult,
};

/// Macro for deriving `struct`s and `impl`s for use with `deno_core`.
/// Please see the [crate-level documentation](.).
#[proc_macro_attribute]
pub fn js(attr: proc_macro::TokenStream, item: proc_macro::TokenStream) -> proc_macro::TokenStream {
    js_item(attr.into(), item.into()).or_error().into()
}

fn js_item(args: TokenStream, item: TokenStream) -> Result<TokenStream> {
    let errors = Error::accumulator();

    let (js, errors) = FlagLike::<JsItem>::parse_macro_attribute(args).or_fatal(errors)?;

    let (js, errors) = match js.0 {
        JsItem::Value(FlagLike(value)) => value::value(value, item).error_at::<JsItem, Value>(),

        JsItem::Module(FlagLike(module)) => {
            module::module(module, item).error_at::<JsItem, Module>()
        }
        JsItem::GlobalThis(FlagLike(global_this)) => {
            global_this::global_this(global_this, item).error_at::<JsItem, GlobalThis>()
        }
        JsItem::Interface(FlagLike(interface)) => {
            interface::interface(interface, item).error_at::<JsItem, Interface>()
        }
        JsItem::Callable(FlagLike(callable)) => {
            callable::callable(callable, item).error_at::<JsItem, Callable>()
        }
        JsItem::Iterator(FlagLike(iterator)) => {
            iterator::iterator(iterator, item).error_at::<JsItem, Iterator_>()
        }
    }
    .or_fatal(errors)?;

    errors.finish_with(js)
}

#[derive(Debug, Clone, FromMeta)]
#[darling(rename_all = "snake_case")]
enum JsItem {
    Module(FlagLike<Module>),
    GlobalThis(FlagLike<GlobalThis>),
    Value(FlagLike<Value>),
    Interface(FlagLike<Interface>),
    Callable(FlagLike<Callable>),
    Iterator(FlagLike<Iterator_>),
}

#[derive(Debug, Clone, FromMeta)]
#[darling(rename_all = "snake_case")]
enum JsProp {
    Prop(FlagLike<Property>),
    Func(FlagLike<Function>),
    New(FlagLike<Constructor>),
    GetIndex(FlagLike<Getter>),
    SetIndex(FlagLike<Setter>),
}

#[derive(Debug, Clone, FromMeta)]
#[doc = include_str!("../../../docs/src/reference/module.md")]
struct Module(Positional<String, ModuleOptions>);

#[derive(Debug, Clone, FromMeta)]
struct ModuleOptions {
    #[darling(default)]
    url: ImportMetaUrl,
    fast: Option<FastString>,
}

#[derive(Debug, Default, Clone)]
enum ImportMetaUrl {
    #[default]
    Preserve,
    Cwd,
    Url(String),
}

#[derive(Debug, Clone, Copy)]
enum FastString {
    Fast,
    FastUnsafeDebug,
}

#[doc = include_str!("../../../docs/src/reference/global-this.md")]
#[derive(Debug, Default, Clone, FromMeta)]
struct GlobalThis;

#[doc = include_str!("../../../docs/src/reference/value.md")]
#[derive(Debug, Default, Clone, FromMeta)]
struct Value {
    #[darling(default)]
    of_type: Unary<V8InnerType>,
}

#[doc = include_str!("../../../docs/src/reference/interface.md")]
#[doc = include_str!("../../../docs/src/reference/_snippets/todo-list.md")]
#[derive(Debug, Default, Clone, FromMeta)]
struct Interface;

#[doc = include_str!("../../../docs/src/reference/callable.md")]
#[derive(Debug, Default, Clone, FromMeta)]
struct Callable;

#[doc = include_str!("../../../docs/src/reference/iterator.md")]
#[derive(Debug, Default, Clone, FromMeta)]
struct Iterator_;

type PropKeyString = StringLike<String>;

type PropKeySymbol = StringLike<WellKnown>;

#[doc = include_str!("../../../docs/src/reference/interface/prop.md")]
#[derive(Debug, Default, Clone, FromMeta)]
struct Property {
    name: Option<Unary<PropKeyString>>,
    #[darling(rename = "Symbol")]
    symbol: Option<Unary<PropKeySymbol>>,
    with_setter: Flag,
}

#[doc = include_str!("../../../docs/src/reference/interface/func.md")]
#[derive(Debug, Default, Clone, FromMeta)]
struct Function {
    name: Option<Unary<PropKeyString>>,
    #[darling(rename = "Symbol")]
    symbol: Option<Unary<PropKeySymbol>>,
}

#[doc = include_str!("../../../docs/src/reference/interface/new.md")]
#[derive(Debug, Default, Clone, FromMeta)]
struct Constructor {
    class: Option<Unary<PropKeyString>>,
}

#[doc = include_str!("../../../docs/src/reference/interface/get-set.md")]
#[derive(Debug, Default, Clone, FromMeta)]
struct Getter;

#[derive(Debug, Default, Clone, FromMeta)]
struct Setter;

impl FromMeta for ImportMetaUrl {
    fn from_list(items: &[NestedMeta]) -> Result<Self> {
        match items.len() {
            0 => Err(Error::too_few_items(1)),
            1 => match &items[0] {
                NestedMeta::Meta(meta) => match path_to_string(meta.path()).as_str() {
                    "preserve" => Ok(Self::Preserve),
                    "cwd" => Ok(Self::Cwd),
                    other => Err(Error::unknown_field_with_alts(
                        other,
                        &["preserve", "cwd", "a string literal"],
                    )),
                },
                NestedMeta::Lit(Lit::Str(lit)) => Ok(Self::Url(lit.value())),
                NestedMeta::Lit(lit) => Err(Error::unexpected_lit_type(lit)),
            },
            _ => Err(Error::too_many_items(1)),
        }
    }
}

impl FromMeta for FastString {
    fn from_list(items: &[NestedMeta]) -> Result<Self> {
        match items.len() {
            0 => Self::from_word(),
            1 => {
                let path = match &items[0] {
                    NestedMeta::Meta(Meta::Path(path)) => path_to_string(path),
                    NestedMeta::Lit(Lit::Str(lit)) => lit.value(),
                    NestedMeta::Meta(Meta::List(_)) => {
                        return Err(Error::unsupported_format("list"))
                    }
                    NestedMeta::Meta(Meta::NameValue(_)) => {
                        return Err(Error::unsupported_format("name value"))
                    }
                    NestedMeta::Lit(lit) => return Err(Error::unexpected_lit_type(lit)),
                };
                match &*path {
                    "unsafe_debug" => Ok(Self::FastUnsafeDebug),
                    path => Err(Error::unknown_field_with_alts(path, &["unsafe_debug"])),
                }
            }
            _ => Err(Error::too_many_items(1)),
        }
    }

    fn from_word() -> Result<Self> {
        Ok(Self::Fast)
    }
}

impl FlagName for JsItem {
    const PREFIX: &'static str = "js";

    fn unit() -> Result<Self> {
        JsItem::from_word()
    }
}

impl FlagEnum for JsItem {
    const PREFIXES: &'static [&'static str] = &[
        Module::PREFIX,
        GlobalThis::PREFIX,
        Value::PREFIX,
        Interface::PREFIX,
        Iterator_::PREFIX,
    ];
}

impl FlagName for JsProp {
    const PREFIX: &'static str = "js";

    fn unit() -> Result<Self> {
        JsProp::from_word()
    }
}

impl FlagEnum for JsProp {
    const PREFIXES: &'static [&'static str] = &[
        Property::PREFIX,
        Function::PREFIX,
        Constructor::PREFIX,
        Getter::PREFIX,
        Setter::PREFIX,
    ];
}

impl FlagName for Module {
    const PREFIX: &'static str = "module";

    fn unit() -> Result<Self> {
        Module::from_word()
    }
}

impl FlagName for GlobalThis {
    const PREFIX: &'static str = "global_this";

    fn unit() -> Result<Self> {
        Ok(Self)
    }
}

impl FlagName for Value {
    const PREFIX: &'static str = "value";

    fn unit() -> Result<Self> {
        Ok(Default::default())
    }
}

impl FlagName for Interface {
    const PREFIX: &'static str = "interface";

    fn unit() -> Result<Self> {
        Ok(Self)
    }
}

impl FlagName for Callable {
    const PREFIX: &'static str = "callable";

    fn unit() -> Result<Self> {
        Ok(Self)
    }
}

impl FlagName for Iterator_ {
    const PREFIX: &'static str = "iterator";

    fn unit() -> Result<Self> {
        Ok(Self)
    }
}

impl FlagName for Property {
    const PREFIX: &'static str = "prop";

    fn unit() -> Result<Self> {
        Ok(Default::default())
    }
}

impl FlagName for Function {
    const PREFIX: &'static str = "func";

    fn unit() -> Result<Self> {
        Ok(Default::default())
    }
}

impl FlagName for Constructor {
    const PREFIX: &'static str = "new";

    fn unit() -> Result<Self> {
        Ok(Default::default())
    }
}

impl FlagName for Getter {
    const PREFIX: &'static str = "get_index";

    fn unit() -> Result<Self> {
        Ok(Self)
    }
}

impl FlagName for Setter {
    const PREFIX: &'static str = "set_index";

    fn unit() -> Result<Self> {
        Ok(Self)
    }
}

#[proc_macro]
pub fn unsafe_include_fast_string(args: proc_macro::TokenStream) -> proc_macro::TokenStream {
    let path = parse_macro_input!(args as LitStr);
    fast_string::unsafe_include_fast_string(path).into()
}
