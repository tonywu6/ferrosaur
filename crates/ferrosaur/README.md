# ferrosaur

<div class="badges">

[![crates.io](https://img.shields.io/crates/v/ferrosaur?style=flat-square)](https://crates.io/crates/ferrosaur)
[![documentation](https://img.shields.io/github/actions/workflow/status/tonywu6/ferrosaur/docs.yml?event=release&style=flat-square&label=docs)](https://tonywu6.github.io/ferrosaur/)
[![MIT/Apache-2.0 licensed](https://img.shields.io/crates/l/ferrosaur?style=flat-square)](https://github.com/tonywu6/ferrosaur/tree/main/LICENSE-APACHE.md)

</div>

_ferrosaur_ derives statically-typed Rust code — à la [wasm-bindgen] — for
[`deno_core::JsRuntime`] to interface with your JavaScript code.

<figure>

If you have:

```js
// lib.js
export const slowFib = (n) =>
  n === 0 ? 0 : n === 1 ? 1 : slowFib(n - 1) + slowFib(n - 2);
```

... and you write:

```rust
// lib.rs
use ferrosaur::js;

#[js(module("lib.js"))]
struct Math;

#[js(interface)]
impl Math {
    #[js(func)]
    fn slow_fib(&self, n: serde<usize>) -> serde<usize> {}
}
```

... then you get:

```rust
// let rt: &mut JsRuntime;
let lib = Math::main_module_init(rt).await?;
let fib = lib.slow_fib(42, rt)?;
assert_eq!(fib, 267914296);
```

</figure>

> [!TIP]
>
> This is like the inverse of [`deno_core::op2`]:
>
> - `#[op2]` gives JavaScript programs easy access to your Rust implementation.
> - _ferrosaur_ gives your Rust program easy access to JavaScript implementations.

## License

This project is released under the
[Apache 2.0 License](https://github.com/tonywu6/ferrosaur/tree/main/LICENSE-APACHE.md)
and the [MIT License](https://github.com/tonywu6/ferrosaur/tree/main/LICENSE-MIT.md).

<!-- prettier-ignore-start -->

[`deno_core::JsRuntime`]: https://docs.rs/deno_core/latest/deno_core/struct.JsRuntime.html
[`deno_core::op2`]:       https://docs.rs/deno_ops/latest/deno_ops/attr.op2.html
[wasm-bindgen]:           https://github.com/rustwasm/wasm-bindgen#example

<!-- prettier-ignore-end -->
