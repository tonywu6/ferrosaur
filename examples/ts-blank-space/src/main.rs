use std::path::Path;

use anyhow::Result;
use bat::PrettyPrinter;
use ferrosaur::js;

use example_runtime::{
    deno,
    deno_core::{self, serde, url::Url},
};
use example_ts::{inject_env_vars, Compiler, TypeScript};

#[js(module("../dist/main.js", fast))]
struct Main;

#[js(interface)]
impl Main {
    #[js(func(name = "default"))]
    fn blank_space<S: serde::Serialize>(&self, src: serde<S>) -> serde<String> {}
}

#[tokio::main]
async fn main() -> Result<()> {
    let rt = &mut deno(Main::module_url()?)?.js_runtime;

    TypeScript::side_module_init(rt).await?;

    inject_env_vars(rt)?;

    let ts = Main::main_module_init(rt).await?;

    let source = Path::new(env!("CARGO_MANIFEST_DIR")).join("../ts/src/lib.ts");

    let js = ts.blank_space(std::fs::read_to_string(&source)?, rt)?;

    #[js(value(of_type(v8::Object)))]
    struct Module;

    let module: Module = {
        let url = Url::from_directory_path(env!("CARGO_MANIFEST_DIR"))
            .unwrap()
            .join("ad-hoc.js")?;
        let id = rt.load_side_es_module_from_code(&url, js.clone()).await?;
        rt.mod_evaluate(id).await?;
        rt.get_module_namespace(id)?.into()
    };

    impl Compiler for Module {}

    PrettyPrinter::new()
        .input_from_bytes(js.as_bytes())
        .language("javascript")
        .theme("GitHub")
        .print()?;

    println!();

    let errors = module
        .create_program(vec![source.to_string_lossy().into()], rt)?
        .print_diagnostics(true, rt)?;

    println!("{errors}");

    {
        let mut settings = insta::Settings::clone_current();
        settings.set_description("script compiled with ts-blank-space");
        settings.set_prepend_module_to_snapshot(false);
        settings.set_snapshot_path("../tests/snapshots");
        settings.bind(|| insta::assert_snapshot!(js));
    }

    Ok(())
}
