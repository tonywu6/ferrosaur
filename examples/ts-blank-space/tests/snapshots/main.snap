---
source: examples/ts-blank-space/src/main.rs
description: script compiled with ts-blank-space
expression: js
---
import ts from "npm:typescript";

export function createProgram(...files          ) {
  const host = createCompilerHost();

  const program = ts.createProgram(
    files,
    {
      strict: true,
      noEmit: true,
      target: ts.ScriptTarget.ESNext,
      lib: ["ESNext"],
      module: ts.ModuleKind.ESNext,
      moduleResolution: ts.ModuleResolutionKind.Bundler,
      moduleDetection: ts.ModuleDetectionKind.Force,
    },
    host,
  );

  return {
    printDiagnostics: (colored = true) => {
      const diagnostics = program.getSemanticDiagnostics();
      if (colored) {
        return ts.formatDiagnosticsWithColorAndContext(diagnostics, diagnosticsHost);
      } else {
        return ts.formatDiagnostics(diagnostics, diagnosticsHost);
      }
    },
  };
}

function createCompilerHost()                  {
  const host = ts.createCompilerHost({});

  const { readFile } = host;

  const runtimeDir = new URL(".", import.meta.url).pathname;

  host.getDefaultLibLocation = () => runtimeDir;

  host.readFile = (fileName) => {
    if (fileName.startsWith(runtimeDir)) {
      const suffix = fileName.slice(runtimeDir.length);
      const content = globalThis.TYPESCRIPT_LIB?.[suffix];
      if (typeof content === "string") {
        return content;
      }
    }
    return readFile(fileName);
  };

  return host;
}

;               
                        
                              
                                                           
                              
                                               
   
 

const diagnosticsHost                           = {
  getCurrentDirectory: () => globalThis.CARGO_MANIFEST_DIR || Deno.cwd(),
  getCanonicalFileName: (name) => name,
  getNewLine: () => "\n",
};
